src/vulkan-renderer/world/collision.cpp

```diff
@@ -0,0 +1,184 @@
+                                                  const glm::vec3 &ray_pos, const glm::vec3 &ray_dir) {
+        return ray_pos - ray_dir * (glm::dot((ray_pos - plane_pos), plane_norm) / glm::dot(ray_dir, plane_norm));
+    };
+
+    /// @brief As soon as we determined which face is selected and we calculated the intersection point between the
+    /// ray and the face plane, we need to determine the nearest corner in that face and the nearest edge on that
+    /// face. In order to determine the nearest corner on a selected face for example, we would iterate through all
+    /// corners on the selected face and calculate the distance between the intersection point and the corner's
+    /// coordinates. The corner which is closest to the intersection point is the selected corner.
+    /// However, we should not use ``glm::distance`` for this, because it performs a ``sqrt`` calculation on the
+    /// vector coordinates. This is not necessary in this case, as we are not interested in the exact distance but
+    /// rather in a value which allows us to determine the nearest corner. This means we can use the squared
+    /// distance, which allows us to avoid the costly call of ``sqrt``.
+    /// @param pos1 The first point.
+    /// @param pos2 The second point.
+    const auto square_of_distance = [&](const std::array<glm::vec3, 2> &points) {
+        return glm::distance2(points[0], points[1]);
+    };
+
+    float shortest_squared_distance{std::numeric_limits<float>::max()};
+    float squared_distance{std::numeric_limits<float>::max()};
+
+    std::size_t selected_face_index{0};
+
+    // Loop though all faces of the cube and check for collision between ray and face plane.
+    for (std::size_t i = 0; i < 6; i++) {
+
+        // Check if the cube side is facing the camera: if the dot product of the two vectors is smaller than
+        // zero, the corresponding angle is smaller than 90 degrees, so the side is facing the camera. Check the
+        // references page for a detailed explanation of this formula.
+        if (glm::dot(cube_directions[i], ray_dir) < 0.0f) {
+
+            const auto intersection =
+                ray_plane_intersection_point(std::get<1>(cube_face_centers[i]), cube_directions[i], ray_pos, ray_dir);
+
+            squared_distance = square_of_distance({m_cube.center(), intersection});
+
+            if (squared_distance < shortest_squared_distance) {
+                selected_face_index = i;
+                shortest_squared_distance = squared_distance;
+                m_intersection = intersection;
+                m_selected_face = cube_face_centers[i];
+            }
+        }
+    }
+
+    // Reset value to maximum for the search of the closest corner.
+    shortest_squared_distance = std::numeric_limits<float>::max();
+
+    // Loop through all corners of this face and check for the nearest one.
+    for (const auto corner_index : CORNERS_ON_FACE_INDICES[selected_face_index]) {
```

Comment:
What happens if two corners are over each other?

Reply:
They can't be. Every face has 4 separate corners.

Reply:
But they can be indented over each other.

Reply:
Those are the corners of the bounding box, not the coordinates of the vertices after indentation. These corner coordinates will stay the same, no matter what indentation takes place.

Reply:
I clarified that this is bounding box data, not directly cube data.